spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindPotentialSecurityCheckBasedOnUntrustedSource.java

```diff
@@ -0,0 +1,242 @@
+package edu.umd.cs.findbugs.detect;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+
+import org.apache.bcel.Const;
+import org.apache.bcel.classfile.Code;
+import org.apache.bcel.classfile.JavaClass;
+import org.apache.bcel.classfile.LocalVariable;
+import org.apache.bcel.classfile.Method;
+
+import edu.umd.cs.findbugs.BugAccumulator;
+import edu.umd.cs.findbugs.BugInstance;
+import edu.umd.cs.findbugs.BugReporter;
+import edu.umd.cs.findbugs.OpcodeStack;
+import edu.umd.cs.findbugs.SourceLineAnnotation;
+import edu.umd.cs.findbugs.ba.AnalysisContext;
+import edu.umd.cs.findbugs.ba.XField;
+import edu.umd.cs.findbugs.ba.XMethod;
+import edu.umd.cs.findbugs.ba.ch.Subtypes2;
+import edu.umd.cs.findbugs.bcel.OpcodeStackDetector;
+import edu.umd.cs.findbugs.classfile.ClassDescriptor;
+
+public class FindPotentialSecurityCheckBasedOnUntrustedSource extends OpcodeStackDetector {
+    private static class CalleeInfo {
+        ClassDescriptor calledClass;
+        XMethod calledMethod;
+        String calledOnName;
+        SourceLineAnnotation srcLine;
+
+        CalleeInfo(ClassDescriptor cls, XMethod called, String name, SourceLineAnnotation line) {
+            calledClass = cls;
+            calledMethod = called;
+            calledOnName = name;
+            srcLine = line;
+        }
+    }
+
+    private static class CallerInfo {
+        OpcodeStack.Item calledOn;
+        JavaClass callerClass;
+        SourceLineAnnotation srcLine;
+
+        CallerInfo(OpcodeStack.Item obj, JavaClass cls, SourceLineAnnotation line) {
+            calledOn = obj;
+            callerClass = cls;
+            srcLine = line;
+        }
+    }
+
+    private static class CallPair {
+        CalleeInfo outside;
+        CallerInfo inside;
+
+        CallPair(CalleeInfo out, CallerInfo in) {
+            outside = out;
+            inside = in;
+        }
+    }
+
+    private Map<XMethod, Set<CalleeInfo>> nonFinalMethodsCalledOnParam = new HashMap<>();
+    private Map<XMethod, Set<CallerInfo>> methodsCalledInsidePrivilegedAction = new HashMap<>();
+
+    private String topParameterName = null;
+
+    private boolean isDoPrivileged = false;
+    private boolean isDoPrivilegedRun = false;
+
+    BugAccumulator bugAccumulator;
+
+    public FindPotentialSecurityCheckBasedOnUntrustedSource(BugReporter bugReporter) {
+        this.bugAccumulator = new BugAccumulator(bugReporter);
+    }
+
+    @Override
+    public void visit(JavaClass obj) {
+        isDoPrivileged = Subtypes2.instanceOf(getDottedClassName(), "java.security.PrivilegedAction")
+                || Subtypes2.instanceOf(getDottedClassName(), "java.security.PrivilegedExceptionAction");
+    }
+
+    @Override
+    public void visit(Method obj) {
+        nonFinalMethodsCalledOnParam.clear();
+        isDoPrivilegedRun = isDoPrivileged && "run".equals(getMethodName());
```

Comment:
better to check the method descriptor (arguments) too. user may add other `run` methods which isn't our target to verify.